File: sim/NationRise.Core/World/WorldState.cs
using NationRise.Core.Determinism;
using NationRise.Core.Time;

namespace NationRise.Core.World;

public sealed class WorldState
{
    public ProvinceStore Provinces { get; }
    public NationStore Nations { get; }
    public GameClock Clock { get; }
    public DeterministicRandom Random { get; }

    public WorldState(
        ProvinceStore provinces,
        NationStore nations,
        GameClock clock,
        DeterministicRandom random)
    {
        Provinces = provinces ?? throw new ArgumentNullException(nameof(provinces));
        Nations = nations ?? throw new ArgumentNullException(nameof(nations));
        Clock = clock ?? throw new ArgumentNullException(nameof(clock));
        Random = random ?? throw new ArgumentNullException(nameof(random));
    }

    public int VictoryPointsOf(ushort nation)
    {
        int points = 0;

        for (int i = 0; i < Provinces.Count; i++)
        {
            if (Provinces.Controller[i] != nation)
            {
                continue;
            }

            points += Provinces.IsCity[i]
                ? (int)MathF.Round(Provinces.Population[i])
                : 1;
        }

        return points;
    }
}

File: web/src/sim/world/WorldState.ts
import type { DeterministicRandom } from '~/sim/determinism/DeterministicRandom'
import { roundHalfToEven } from '~/sim/determinism/rounding'
import type { GameClock } from '~/sim/time/GameClock'
import type { NationStore } from '~/sim/world/NationStore'
import type { ProvinceStore } from '~/sim/world/ProvinceStore'

export class WorldState {
  constructor(
    readonly provinces: ProvinceStore,
    readonly nations: NationStore,
    readonly clock: GameClock,
    readonly random: DeterministicRandom,
  ) {}

  victoryPointsOf(nation: number): number {
    let points = 0

    for (let i = 0; i < this.provinces.count; i++) {
      if (this.provinces.controller[i] !== nation) continue

      points += this.provinces.isCity[i] !== 0
        ? roundHalfToEven(this.provinces.population[i]!)
        : 1
    }

    return points
  }
}
